### PR TITLE
Launcher refactoring

### PR DESCRIPTION
**Pull request by pitr-ch**

### CODE REVIEW

File: mx.truffleruby/suite.py [hunk 1]
```diff
@@ -30,33 +30,33 @@
...
-          "url": "http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html"
+            "name": "GNU General Public License, version 2",
+            "url": "http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html"
         },
         "zlib": {
-          "name": "The zlib License",
-          "url": "https://opensource.org/licenses/zlib"
+            "name": "The zlib License",
+            "url": "https://opensource.org/licenses/zlib"
         },
```
Comment (chrisseaton) [inline mx.truffleruby/suite.py]: Why so many whitespace changes? Here and later on as well.

File: mx.truffleruby/suite.py [hunk 2]
```diff
@@ -296,9 +296,9 @@
 
         # Set of extra files to extract to run Ruby
         "TRUFFLERUBY-ZIP": {
-            "native": True, # Not Java
+            "native": True,  # Not Java
             "relpath": True,
-            "platformDependent": True, # truffleruby-cext
+            "platformDependent": True,  # truffleruby-cext
```
Comment (chrisseaton) [inline mx.truffleruby/suite.py]: Why these whitespace changes?

File: spec/truffle/launcher_spec.rb [hunk 3]
```diff
@@ -85,31 +85,31 @@ def should_print_full_java_command(cmd)
   end
 
   it "takes options from system properties set in JAVA_OPTS" do
-    out = `JAVA_OPTS=-Dpolyglot.ruby.verbosity=2 #{RbConfig.ruby} -e 'puts $VERBOSE'`
+    out = `JAVA_OPTS=-Dpolyglot.ruby.verbosity=true #{RbConfig.ruby} -e 'puts $VERBOSE'`
```
Comment (chrisseaton) [inline spec/truffle/launcher_spec.rb]: Why this change?

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java [hunk 4]
```diff
@@ -32,70 +32,57 @@
...
+
+        if (current == null) {
+            builder = new StringBuilder();
+            setOptionRaw(key, builder);
+        } else if (current instanceof StringBuilder) {
+            builder = (StringBuilder) current;
+        } else {
+            builder = new StringBuilder(current);
+            setOptionRaw(key, builder);
+        }
```
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: Isn't StringBuilder a premature optimization introducing extra complexity?

File: src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java [hunk 5]
```diff
@@ -48,14 +40,34 @@
...
+            "ruby.show_version", 
+            "Show version.", 
+            false);
+    public static final BooleanOptionDescription SHOW_COPYRIGHT = new BooleanOptionDescription(
+            "ruby.show_copyright", 
+            "Show copyright.", 
+            false);
+    public static final ShowHelpOptionDescription SHOW_HELP = new ShowHelpOptionDescription(
+            "ruby.show_help", 
+            "WHat help should be shown.",
```
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java]: What

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 6]
```diff
@@ -88,52 +89,119 @@ public DynamicObject forceContext() {
         }
     }
 
-    @CoreMethod(names = "main", onSingleton = true, optional = 1)
+    @CoreMethod(names = "main", onSingleton = true)
     public abstract static class MainNode extends CoreMethodArrayArgumentsNode {
 
+        @Child private SnippetNode findSFileSnippetNode;
+        @Child private SnippetNode checkSyntaxSnippetNode;
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: These could be in `@Cached` since there is a single specialization (and even if there was not since this node is executed once).

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 7]
```diff
@@ -238,9 +299,13 @@ public DynamicObject innerCheckSyntax(DynamicObject path) {
         @Specialization(guards = "isRubyString(optionName)")
         public Object getOption(DynamicObject optionName) {
             final OptionDescription<?> description = OptionsCatalog.fromName("ruby." + StringOperations.getString(optionName));
+            assert description != null;
             final Object value = getContext().getOptions().fromDescription(description);
+
             if (value instanceof String) {
                 return makeStringNode.executeMake(value, UTF8Encoding.INSTANCE, CodeRange.CR_UNKNOWN);
+            } else if(value instanceof Enum) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: Missing space.

File: mx.truffleruby/suite.py (hunk 1 above)
Comment (pitr-ch) [inline mx.truffleruby/suite.py]: It's added to unify the indentation in the file. And to align it with idea auto-formatter.

File: mx.truffleruby/suite.py (hunk 2 above)
Comment (pitr-ch) [inline mx.truffleruby/suite.py]: Added with the indentation changes, to match idea auto-formatter.

File: spec/truffle/launcher_spec.rb (hunk 3 above)
Comment (pitr-ch) [inline spec/truffle/launcher_spec.rb]: To make the option more precise so only the enum values are allowed `nil,false,true` as in other enum options.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java [hunk 8]
```diff
@@ -247,7 +247,11 @@ private void processArgument() throws CommandLineException {
                         if (path.startsWith("~" + File.separator)) {
                             path = System.getProperty("user.home") + File.separator + path.substring(2);
                         }
-                        config.getLoadPaths().add(path);
+                        String[] currentPaths = config.getOption(OptionsCatalog.LOAD_PATHS);
+                        String[] newPaths = new String[currentPaths.length + 1];
+                        System.arraycopy(currentPaths, 0, newPaths, 0, currentPaths.length);
```
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java]: good find, but already replaced.

File: spec/truffle/launcher_spec.rb (hunk 3 above)
Comment (chrisseaton) [inline spec/truffle/launcher_spec.rb]: We've lost a little bit of functionality then, as before people who were using system properties to set this option could use the same syntax as they could on the command line. But it doesn't matter that much.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java [hunk 9]
```diff
@@ -32,70 +32,57 @@
...
-            return getScriptFileName();
-        }
+    Map<String, CharSequence> getOptionsInternal() {
+        return options;
     }
 
     public Map<String, String> getOptions() {
-        return options;
+        return options.entrySet().stream().
+                collect(Collectors.toMap(Map.Entry::getKey, (e) -> e.getValue().toString()));
```
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: How does `return new HashMap(options)` convert the values from CharSequence to String?

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java (hunk 4 above)
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: Maybe, but it was there already so I did not want to just drop it. In the end I think it's quite encapsulated and it's not too bad.

File: src/launcher/java/org/truffleruby/launcher/Launcher.java [hunk 10]
```diff
@@ -173,8 +164,13 @@ public static void processArguments(CommandLineOptions config, String[] argument
...
 
-        if (!config.doesHaveScriptToRun() && !config.shouldUsePathScript() && System.console() != null) {
-            config.setUsePathScript("irb");
+        if (config.getOption(OptionsCatalog.EXECUTION_ACTION) == ExecutionAction.UNSET) {
+            if (System.console() != null) {
+                config.setOption(OptionsCatalog.EXECUTION_ACTION, ExecutionAction.PATH);
+                config.setOption(OptionsCatalog.TO_EXECUTE, "irb");
+            } else {
+                config.setOption(OptionsCatalog.EXECUTION_ACTION, ExecutionAction.STDIN);
+            }
```
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/Launcher.java]: I expect to move this around a bit to make it more sharable with the GVM launcher. I already call it from both.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java (hunk 6 above)
Comment (pitr-ch) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: Yeah, could be, but the snippets are not always used which makes this tiny better. I don't think we would gain anything by this change. I think it's good either way so I keep it as is.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java (hunk 4 above)
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: It makes this weird CharSequence signature and needs copying at the high-level, I think it's highly premature, few options are appended and very few times, isn't it?

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java (hunk 6 above)
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: It's 10 lines of codes versus 2, how is that better?
I call for consistency here (and again premature optimization).

File: mx.truffleruby/suite.py (hunk 2 above)
Comment (eregon) [inline mx.truffleruby/suite.py]: Why does it wants 2 spaces before a comment? That doesn't make sense.

**Comment by pitr-ch** [comment]
I've tried either make small commit so the message should be self-explanatory or add bullet-points to explain the changes (in the last big commit). In general I wanted to clean up this code so it's more understandable: explicit enums rather than complicated interactions between group of boolean flags, storing options in one map, work with values rather than strings in the option parser, etc. Now I can add mapping between Ruby options and its polyglot format.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java [hunk 11]
```diff
@@ -54,6 +53,10 @@
     private boolean forceStdin;
     private boolean shouldPrintShortUsage;
 
+    // TODO (pitr-ch 26-Jul-2017): Move to Options when implementing -s option
+    // Currently not used
```
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: I would like to implement this on our side soon, since we already have the parsing in `processArgv` method. Hm I could delete it and keep a hash of a commit to revert around in the comment though.

Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: I would just re-add it manually, it's literally a field declaration and a getter, right?

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java (hunk 4 above)
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: Where is 'weird CharSequence signature' ? `CharSequence` is contained in `CommandLineOptions`.
Where 'needs copying'? Do you mean in the `getOptions`? The coping could be fixed with:
```java
    @SuppressWarnings("unchecked")
    public Map<String, String> freezeAndGetOptions() {
        for (Map.Entry<String, CharSequence> entry : options.entrySet()) {
            entry.setValue(entry.getValue().toString());
        }
        // freeze map so it cannot become none-String again
        options = Collections.unmodifiableMap(options);
        return (Map<String, String>)(Map)options;
    }
```
Anyway, there is no strong reason to keep it. I'll replace it with simple String concatenation.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java (hunk 11 above)
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: No there is also the parsing logic in the processArgv methods, but I've already implemented it so it does not stay behind.

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java (hunk 4 above)
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineOptions.java]: When I see `CharSequence` I immediately wonder "Why is this not String, and what hack is it hiding"? In this case, it hides that the values of the map are both String or mutable StringBuffer in a mutable Map, which seems like a lot of complexity.
Reading further, I see this very-hard to read Stream expression, where it could be a simple and easy-to-read copying loop. Trying to be functional here is paradoxal to this CharSequence design in the first place. Cheating the type system and giving up type safety is not the way to go either.
I don't mind creating a new HashMap at all, but this whole CharSequence thing to optimize a few StringBuilder allocations really does not seem worth the induced complexity and the time needed to read and understand this code (which is why I get a bit rough in my comments here).
Simple code is easier to read, review and refactor.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 12]
```diff
@@ -88,52 +89,119 @@ public DynamicObject forceContext() {
...
+                                this,
+                                to_execute);
+                        dollarZeroValue = StringOperations.createString(getContext(),
+                                StringOperations.encodeRope("-", UTF8Encoding.INSTANCE));
+                        break;
+
+                    case INLINE:
+                        source = getContext().getSourceLoader().loadMainEval();
+                        dollarZeroValue = StringOperations.createString(getContext(),
+                                StringOperations.encodeRope("-", UTF8Encoding.INSTANCE));
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: This should be `"-e"`: `ruby -e 'p $0'`

File: mx.truffleruby/suite.py (hunk 2 above)
Comment (pitr-ch) [inline mx.truffleruby/suite.py]: Ah did not notice, I though it's missing a space. Found a way to reconfigure the autformatter, I'll fix

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java [hunk 13]
```diff
@@ -156,17 +155,26 @@ private void processArgv() {
...
+                    value = arg.substring(split + 1);
                 } else {
-                    config.getOptionGlobals().put(arg, null);
+                    key = arg;
+                    value = null;
+                }
+
+                final StringArrayOptionDescription optionDescription =
+                        value != null ? OptionsCatalog.ARGV_GLOBAL_VALUES : OptionsCatalog.ARGV_GLOBAL_FLAGS;
+                // argv globals getService their dashes replaced with underscores
```
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java]: What is this `getService` mentioned here?

File: src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java [hunk 14]
```diff
@@ -76,6 +76,18 @@
...
+            "Parse options in script argv into global variables.", 
+            false);
+    public static final StringArrayOptionDescription ARGV_GLOBAL_VALUES = new StringArrayOptionDescription(
+            "ruby.argv_global_values", 
+            "Parsed options from script argv with a value.", 
+            new String[]{});
+    public static final StringArrayOptionDescription ARGV_GLOBAL_FLAGS = new StringArrayOptionDescription(
+            "ruby.argv_global_flags", 
+            "Parsed options from script argv acting as flags (no value).", 
+            new String[]{});
```
Comment (eregon) [inline src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java]: Are the flags not the same as if they had the value `true`?
The problem is then differentiating the boolean true and the String "true" since both have to be passed as String in options?

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 15]
```diff
@@ -131,6 +133,28 @@ public int main(VirtualFrame frame, @Cached("create()") IndirectCallNode callNod
...
+                assert global_values.length % 2 == 0;
+                for (int i = 0; i < global_values.length; i += 2) {
+                    String key = global_values[i];
+                    String value = global_values[i + 1];
+
+                    getContext().getCoreLibrary().getGlobalVariables().put(
+                            "$" + key,
+                            StringOperations.createString(
+                                    getContext(),
+                                    StringOperations.encodeRope(value, UTF8Encoding.INSTANCE)));
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: I think @nirvdrum might want a MakeStringNode here.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 16]
```diff
@@ -92,23 +91,23 @@ public DynamicObject forceContext() {
...
                 // EXECUTION_ACTION was set to NONE
                 return 0;
             }
 
             if (getContext().getOptions().SYNTAX_CHECK) {
-                return (int) getCheckSyntaxSnippetNode().execute(
+                return (int) checkSyntaxSnippetNode.execute(
                         frame,
                         "Truffle::Boot.check_syntax source",
                         "source", source);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: FWIW I prefer a CallDispatchHeadNode rather than a Snippet when it's just one call.
It also does not need the frame if it's created with `CallDispatchHeadNode.createOnSelf()`
as then it does not need to check visibility.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 17]
```diff
@@ -177,7 +176,7 @@ private Source loadMainSourceSettingDollarZero(VirtualFrame frame) {
                         break;
 
                     case PATH:
-                        final DynamicObject path = (DynamicObject) getFindSFileSnippetNode().execute(
+                        final DynamicObject path = (DynamicObject) findSFileSnippetNode.execute(
                                 frame,
                                 "Truffle::Boot.find_s_file Truffle::Boot.get_option('to_execute')");
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: This is actually also just one call since the option could be trivially accessed from Java.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java (hunk 16 above)
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: Removing the need for the VritualFrame is important here as this code could easily no longer PE, and it should be a TruffleBoundary.

File: src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java (hunk 14 above)
Comment (pitr-ch) [inline src/launcher/java/org/truffleruby/launcher/options/OptionsCatalog.java]: Yeah, there would not be a way to distinguish between a switch without a value (which becomes `true`) and a one with a value `"true"`. I'll add a comment.

File: src/main/ruby/core/truffle/boot.rb [hunk 18]
```diff
@@ -27,6 +27,8 @@ def self.check_syntax(source_or_file)
     1
   end
 
+  private_class_method :check_syntax
```
Comment (eregon) [inline src/main/ruby/core/truffle/boot.rb]: The convention for `private` and friends (in our codebase) is to add them right after the `end` of the method they apply to, so it makes a logical group of what belongs together.
Alternatively `private def my_method` works too but here with `private_class_method` it would be very long and not so readable.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 19]
```diff
@@ -94,23 +96,25 @@ public DynamicObject forceContext() {
...
-                @Cached("new()") SnippetNode checkSyntaxSnippetNode) {
+                @Cached("createOnSelf()") CallDispatchHeadNode findSFile,
+                @Cached("createOnSelf()") CallDispatchHeadNode checkSyntax,
+                @Cached("create()") StringNodes.MakeStringNode makeStringNode) {
 
-            setArgvGlobals();
+            setArgvGlobals(makeStringNode);
 
-            Source source = loadMainSourceSettingDollarZero(frame, findSFileSnippetNode);
+            Source source = loadMainSourceSettingDollarZero(frame, findSFile, makeStringNode);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: When using `createOnSelf()`, you can pass `null` for the `frame`. Therefore the whole specialization can be `@TruffleBoundary` which seems safer here given the amount of code and branches.

File: src/main/ruby/core/truffle/boot.rb (hunk 18 above)
Comment (pitr-ch) [inline src/main/ruby/core/truffle/boot.rb]: When scanning our source code, having a space seems more frequent. I like the space better. Auto-formatter inserts the space as well and personally I have no desire to fight the auto-formatting. I would not mind `private_class_method def ...` pattern thought, but I think that's topic for another Rubocop rules adoption PR if there is a rule for it. I am inclined to keep the space.

File: src/main/java/org/truffleruby/language/TruffleBootNodes.java [hunk 20]
```diff
@@ -209,7 +209,7 @@ private Source loadMainSourceSettingDollarZero(
                 throw new JavaException(e);
             }
 
-            getContext().getCoreLibrary().getGlobalVariables().put("$0", dollarZeroValue);
+            getContext().getCoreLibrary().getGlobalVariables().getStorage("$0").setValueInternal(dollarZeroValue);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/TruffleBootNodes.java]: I think we should rename `put` to `register` or so to emphase it's only for declaring new global variables (the check is useful as during startup we don't want to blindly override). But that's for another PR.

File: src/main/ruby/core/truffle/boot.rb (hunk 18 above)
Comment (eregon) [inline src/main/ruby/core/truffle/boot.rb]: RubyMine's formatter seems to have rather strange defaults. Fine, let's keep it like that for now.
